File: normal_order_java/79.WordSearch.java
class Solution {
    int columns, rows;
    boolean ans;
    String word;
    public boolean exist(char[][] board, String word) {
        // if(board == null || board[0] == null) return false;
        this.ans = false;
        this.word = word;
        this.rows = board.length;
        this.columns = board[0].length;
        for(int row = 0; row < rows; row++){
            for(int column = 0; column < columns; column++){
                if(board[row][column] == word.charAt(0) && ans == false){
                    board[row][column] = '#';       
                    dfs(board, 1, row, column);
                    board[row][column] = word.charAt(0);
                    // System.out.println();
                }
            }
        }
        return ans;
    }
    
    public void dfs(char[][] board, int index, int r, int c){
        if(index == word.length()){
            ans = true;
            return;
        }
        int[] dr = {0, 0, 1, -1};
        int[] dc = {1, -1, 0, 0};
        for(int i = 0; i<=3; i++ ){
            int x = r+dr[i], y = c+dc[i];
            if(0<=x && x<rows && 0<=y && y<columns && board[x][y] == word.charAt(index) && ans == false){
                board[x][y] = '#';                    
                dfs(board, index+1, x, y);
                board[x][y] = word.charAt(index);                
            }
        }
    }
}

class Solution {
    int columns, rows;
    boolean ans;
    String word;
    static boolean[][] visited;
    public boolean exist(char[][] board, String word) {
        // if(board == null || board[0] == null) return false;
        visited = new boolean[board.length][board[0].length];
        this.ans = false;
        this.word = word;
        this.rows = board.length;
        this.columns = board[0].length;
        for(int row = 0; row < rows; row++){
            for(int column = 0; column < columns; column++){
                if(board[row][column] == word.charAt(0) && ans == false){
                    // System.out.print(word.charAt(0));
                    visited[row][column] = true;
                    dfs(board, 1, row, column);
                    visited[row][column] = false;
                    // System.out.println();
                }
            }
        }
        return ans;
    }
    
    public void dfs(char[][] board, int index, int r, int c){
        if(index == word.length()){
            ans = true;
            return;
        }
        int[] dr = {0, 0, 1, -1};
        int[] dc = {1, -1, 0, 0};
        for(int i = 0; i<=3; i++ ){
            int x = r+dr[i], y = c+dc[i];
            if(0<=x && x<rows && 0<=y && y<columns && board[x][y] == word.charAt(index) && ans == false && visited[x][y] == false){
                // System.out.print(word.charAt(index));
                visited[x][y] = true;
                dfs(board, index+1, x, y);
                visited[x][y] = false;               
            }
        }
    }
}

public class WordSearch{
    public static void main(String[] args){
        char[][] board = {{'b','a','a','b','a','b'},{'a','b','a','a','a','a'},{'a','b','a','a','a','b'},{'a','b','a','b','b','a'},{'a','a','b','b','a','b'},{'a','a','b','b','b','a'},{'a','a','b','a','a','b'}};
        String word = "ababaababaaabbabbaabbaabbaba";
        Solution s = new Solution();
        System.out.println(s.exist(board, word));
    }
}

File: normal_order_python/79.WordSearch.py
'''
tags: backtrack(DFS)

79. Word Search

Medium

Given a 2D board and a word, find if the word exists in the grid.

The word can be constructed from letters of sequentially adjacent cell, where "adjacent" cells are those horizontally or vertically neighboring. The same letter cell may not be used more than once.

Example:

board =
[
  ['A','B','C','E'],
  ['S','F','C','S'],
  ['A','D','E','E']
]

Given word = "ABCCED", return true.
Given word = "SEE", return true.
Given word = "ABCB", return false.
'''
class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        self.ans = False
        self.word = word
        self.length = len(word)
        if not board:
            return self.ans
        self.rows, self.columns = len(board), len(board[0])
        for row in range(self.rows):
            for column in range(self.columns):     
                if board[row][column] == word[0] and not self.ans:
                    print(self.word[0], end='')
                    prev = board[row][column]
                    board[row][column] = 'X'                    
                    self.helper(board, 1, row, column)
                    board[row][column] = prev
                    print()
                    
        return self.ans
    
    def helper(self, board, index, r, c):
        if index == self.length:
            self.ans = True
            return
        adjacent = [(r,c+1), (r,c-1), (r+1,c), (r-1,c)]
        for (row, column) in adjacent:
            if 0<=row<self.rows and 0<=column<self.columns and board[row][column] == self.word[index]:
                if not self.ans:
                    print(self.word[index], end='')
                    prev = board[r][c] # this matters
                    board[r][c] = 'X'                     
                    self.helper(board, index+1, row, column)
                    board[r][c] = prev